connectors/sources/google_drive.py

```diff
@@ -195,22 +206,234 @@ async def _execute_api_call(self, resource, method, call_api_func, kwargs):
+
+    def __init__(self, json_credentials):
+        """Initialize the GoogleApiClient superclass.
+
+        Args:
+            json_credentials (dict): Service account credentials json.
+        """
+        super().__init__(
+            json_credentials=json_credentials,
+            api_name="drive",
+            api_version="v3",
+            scopes=["https://www.googleapis.com/auth/drive.readonly"],
+        )
+
+    async def ping(self):
+        return await self.api_call(resource="about", method="get", fields="kind")
+
+    async def get_drives(self):
+        """Fetch all shared drive (id, name) from Google Drive
+
+        Yields:
+            dict: Shared drive metadata.
+        """
+
+        async for drive in self.api_call_paged(
+            resource="drives",
+            method="list",
+            fields="nextPageToken,drives(id,name)",
+            pageSize=100,
+        ):
+            yield drive
+
+    async def retrieve_all_drives(self):
+        """Retrieves all shared drives from Google Drive
+
+        Returns:
+            dict: mapping between drive id and its name
+        """
+        drives = {}
+        async for page in self.get_drives():
+            drives_chunk = page.get("drives", [])
+            for drive in drives_chunk:
+                drives[drive["id"]] = drive["name"]
+
+        return drives
+
+    async def get_folders(self):
+        """Fetch all folders (id, name, parent) from Google Drive
+
+        Yields:
+            dict: Folder metadata.
+        """
+        async for folder in self.api_call_paged(
+            resource="files",
+            method="list",
+            corpora="allDrives",
+            fields="nextPageToken,files(id,name,parents)",
+            q=f"mimeType='{FOLDER_MIME_TYPE}' and trashed=false",
+            includeItemsFromAllDrives=True,
+            supportsAllDrives=True,
+            pageSize=1000,
+        ):
+            yield folder
+
+    async def retrieve_all_folders(self):
+        """Retrieves all folders from Google Drive
+
+        Returns:
+            dict: mapping between folder id and its (name, parents)
+        """
+        folders = {}
+        async for page in self.get_folders():
+            folders_chunk = page.get("files", [])
+            for folder in folders_chunk:
+                folders[folder["id"]] = {
+                    "name": folder["name"],
+                    "parents": folder.get("parents", None),
+                }
+
+        return folders
+
+    async def list_files(self):
+        """Get files from Google Drive. Files can have any type.
+
+        Yields:
+            dict: Documents from Google Drive.
+        """
+        async for file in self.api_call_paged(
+            resource="files",
+            method="list",
+            corpora="allDrives",
+            q="trashed=false",
+            orderBy="modifiedTime desc",
+            fields=f"files({DRIVE_ITEMS_FIELDS}),incompleteSearch,nextPageToken",
```

Comment:
Not sure how much additional time is required for returning `permissions` for a Google Drive connector with DLS disabled. Maybe there's no processing in the remote server and just returns an additional field.

Reply:
Ahh good catch, we should try to be efficient and only request `permissions` if dls is disabled 👍